### ISSUE TITLE
Warn if a non-nullable field is used before it is set in the constructor

### ISSUE THREAD

**Issue by BertanAygun** [dotnet/roslyn#71437]
### Describe the problem you are trying to solve

In the code below, `this.notNullString` is marked as not-nullable but is used when value could be null. No warnings were raised since argument in Foo constructor was nullable but in practice code intent was for `notNullString` to be not null when used. A warning here could prevent subtle runtime issues where "test" should not be null in the context of the application it is running under.

```csharp

public class Bar
{
    private readonly string notNullString;

    public Bar(string notNullable)
    {
        new Foo(this.notNullString);
    }
}

public class Foo
{
    public Foo(string? test) { }
}
```

### Describe suggestions on how to achieve the rule

Analyzer would likely have to determine if `notNullString` is initialized before its use site in the constructor.

**Comment by jaredpar** [comment]
> In the code below, this.notNullString is marked as not-nullable but is used when value could be null

That is correct but the compiler recognizes that the value is potentially null before it is assigned. The code in this repo issues no warning because the value is passed to a `string?`. If you change the signature to be `Foo(string test)` the compiler will warn because it recognizes that `new Foo(this.notNullString)` is potentially `null` at that point.
